src/goto-programs/replace_function_bodies.cpp

```diff
@@ -0,0 +1,462 @@
+  }
+
+  if(options == "assume-false")
+  {
+    return util_make_unique<assume_false_replace_function_bodiest>();
+  }
+
+  if(options == "assert-false")
+  {
+    return util_make_unique<assert_false_replace_function_bodiest>();
+  }
+
+  if(options == "assert-false-assume-false")
+  {
+    return util_make_unique<
+      assert_false_then_assume_false_replace_function_bodiest>();
+  }
+
+  if(options == "empty")
+  {
+    return util_make_unique<empty_replace_function_bodiest>();
+  }
+
+  const std::vector<std::string> option_components = split_string(options, ',');
+  if(!option_components.empty() && option_components[0] == "havoc")
+  {
+    std::regex globals_regex;
+    std::regex params_regex;
+    for(std::size_t i = 1; i < option_components.size(); ++i)
+    {
+      const std::vector<std::string> key_value_pair =
+        split_string(option_components[i], ':');
+      if(key_value_pair.size() != 2)
+      {
+        throw replace_function_bodies_errort(
+          "Expected key_value_pair of form argument:value");
+      }
+      if(key_value_pair[0] == "globals")
+      {
+        globals_regex = key_value_pair[1];
+      }
+      else if(key_value_pair[0] == "params")
+      {
+        params_regex = key_value_pair[1];
+      }
+      else
+      {
+        throw replace_function_bodies_errort(
+          "Unknown option \"" + key_value_pair[0] + "\"");
+      }
+    }
+    std::vector<irep_idt> globals_to_havoc;
+    namespacet ns(symbol_table);
+    messaget messages(message_handler);
+    for(auto const &symbol : symbol_table.symbols)
+    {
+      if(
+        symbol.second.is_lvalue && symbol.second.is_static_lifetime &&
+        std::regex_match(id2string(symbol.first), globals_regex))
+      {
+        if(!is_constant_or_has_constant_components(symbol.second.type, ns))
+        {
+          globals_to_havoc.push_back(symbol.first);
+        }
+        else if(!symbol.second.type.get_bool(ID_C_constant))
+        {
+          messages.warning() << "warning: not havoccing non-const global '"
+                             << symbol.first << "' because it has const members"
+                             << messaget::eom;
+        }
+      }
+    }
+    return util_make_unique<havoc_replace_function_bodiest>(
+      std::move(globals_to_havoc), std::move(params_regex), message_handler);
+  }
+  throw replace_function_bodies_errort("Can't parse \"" + options + "\"");
+}
+
+/// Produce a replace_function_body_optionst from a cmdline.
+/// \see replace_function_bodies() for information of how the cmdline is parsed
+replace_function_body_optionst parse_replace_function_body_optionst(
+  const cmdlinet &cmdline,
+  const symbol_tablet &symbol_table,
+  message_handlert &message_handler)
+{
+  if(
+    cmdline.isset("replace-function-body") &&
+    cmdline.isset("generate-function-body"))
+  {
+    throw replace_function_bodies_errort(
+      "can only use one of --replace-function-body "
+      "or --generate-function-body");
```

Comment:
@hannes-steffenhagen-diffblue @peterschrammel - did you discuss this? What was the conclusion, or maybe why was my proposal not acceptable?

Reply:
@tautschnig Nobody else said anything about this. It's possible that they missed it. I wouldn't mind implementing it that way, but it wasn't obviously better than the current behaviour from my perspective, I'd prefer at least one other person saying they think it's the better way to do it before going ahead with changing the options at this point.

Reply:
Poking @martin-cs @chrisr-diffblue with the hope that more people might chime in. Everyone: see the discussion in https://github.com/diffblue/cbmc/pull/1889#discussion_r175534313 for context.